connector/connect/common/src/main/scala/org/apache/spark/sql/connect/client/GrpcExceptionConverter.scala

```diff
@@ -24,46 +24,174 @@ import scala.reflect.ClassTag
           iter.next()
         }
       }
 
       override def close(): Unit = {
-        convert {
+        convert(sessionId, userContext) {
           iter.close()
         }
       }
     }
   }
 
+  private def exceptionInfoToError(exceptionInfo: ExceptionInfo): Option[Error] = {
+    val stackTraceOpt =
+      if (exceptionInfo.getStackTraceList.isEmpty) None
+      else {
+        Some(exceptionInfo.getStackTraceList.asScala.toArray.map { stackTraceElement =>
+          new StackTraceElement(
+            stackTraceElement.getDeclaringClass,
+            stackTraceElement.getMethodName,
+            stackTraceElement.getFileName,
+            stackTraceElement.getLineNumber)
+        })
+      }
+    Some(
+      Error(
+        classHierarchy = exceptionInfo.getErrorTypeHierarchyList.asScala.toArray,
+        message = exceptionInfo.getMessage,
+        stackTraceOpt = stackTraceOpt,
+        causeOpt = if (exceptionInfo.hasCause) {
+          exceptionInfoToError(exceptionInfo.getCause)
+        } else {
+          None
+        }))
+  }
+
+  /**
+   * enrichError enrich errors with full exception message and optionally stacktrace by issuing an
+   * additional RPC call to fetch error details. The RPC call is best-effort at-most-once.
+   */
+  private def enrichError(
+      error: Error,
+      info: ErrorInfo,
+      sessionId: String,
+      userContext: UserContext): Error = {
+    val errorId = info.getMetadataOrDefault("errorId", null)
+    if (errorId == null) {
+      return error
+    }
+
+    try {
+      val errorDetailsResponse = grpcStub.fetchErrorDetails(
+        FetchErrorDetailsRequest
+          .newBuilder()
+          .setSessionId(sessionId)
+          .setErrorId(errorId)
+          .setUserContext(UserContext.newBuilder().setUserId(userContext.getUserId).build())
+          .build())
+
+      if (!errorDetailsResponse.hasExceptionInfo) {
+        return error
+      }
+
+      exceptionInfoToError(errorDetailsResponse.getExceptionInfo).get
+    } catch {
+      case _: StatusRuntimeException => error
+    }
+  }
+
+  private def errorInfoToThrowable(
+      info: ErrorInfo,
+      truncatedMessage: String,
+      sessionId: String,
+      userContext: UserContext): Option[Throwable] = {
+    val classes = deserializeClasses(info.getMetadataOrDefault("classes", "[]"))
+
+    implicit val formats = DefaultFormats
+    var causeOpt: Option[Error] = None
+    JsonMethods
+      .parse(info.getMetadataOrDefault("causes", "[]"))
+      .extract[Array[Map[String, String]]]
+      .reverse
+      .foreach { m =>
+        causeOpt = Some(
+          Error(
+            classHierarchy = deserializeClasses(m.get("classes").get),
+            message = m.get("message").get,
+            stackTraceOpt = None,
+            causeOpt = causeOpt))
+      }
+    val error = Error(classes, truncatedMessage, stackTraceOpt = None, causeOpt = causeOpt)
+    errorToThrowable(enrichError(error, info, sessionId, userContext))
```

Comment:
Why not first try to enrich and then do the fallback. Now we are duplicate work. And, more importantly, it is much harder to follow, a casual reader might get confused.